pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -166,67 +179,109 @@ func BuildDefaultAPIServer(options configapi.MasterConfig) (*apiserveroptions.Se
+	etcdConfig.KeyFile = masterConfig.EtcdClientInfo.ClientCert.KeyFile
+	etcdConfig.CertFile = masterConfig.EtcdClientInfo.ClientCert.CertFile
+	etcdConfig.CAFile = masterConfig.EtcdClientInfo.CA
+
+	storageFactory, err := kubeapiserver.NewStorageFactory(
 		etcdConfig,
-		server.GenericServerRunOptions.DefaultStorageMediaType,
+		server.Etcd.DefaultStorageMediaType,
 		kapi.Codecs,
-		genericapiserver.NewDefaultResourceEncodingConfig(),
+		// FIXME: is this supposed to be resourceEncodingConfig???
+		apiserverstorage.NewDefaultResourceEncodingConfig(kapi.Registry),
 		storageGroupsToEncodingVersion,
 		// FIXME: this GroupVersionResource override should be configurable
 		[]schema.GroupVersionResource{batch.Resource("cronjobs").WithVersion("v2alpha1")},
-		master.DefaultAPIResourceConfigSource(), server.GenericServerRunOptions.RuntimeConfig,
+		master.DefaultAPIResourceConfigSource(),
+		server.APIEnablement.RuntimeConfig,
 	)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	/*storageFactory := genericapiserver.NewDefaultStorageFactory(
-		etcdConfig,
-		server.DefaultStorageMediaType,
-		kapi.Codecs,
-		resourceEncodingConfig,
-		master.DefaultAPIResourceConfigSource(),
-	)*/
 	// the order here is important, it defines which version will be used for storage
 	storageFactory.AddCohabitatingResources(batch.Resource("jobs"), extensions.Resource("jobs"))
 	storageFactory.AddCohabitatingResources(extensions.Resource("horizontalpodautoscalers"), autoscaling.Resource("horizontalpodautoscalers"))
 
-	return server, storageFactory, nil
+	return storageFactory, nil
 }
 
-// TODO this function's parameters need to be refactored
-func BuildKubernetesMasterConfig(
-	options configapi.MasterConfig,
-	requestContextMapper apirequest.RequestContextMapper,
-	kubeClient kclientset.Interface,
-	informers shared.InformerFactory,
-	admissionControl admission.Interface,
-	originAuthenticator authenticator.Request,
-	kubeAuthorizer authorizer.Authorizer,
-) (*MasterConfig, error) {
-	if options.KubernetesMasterConfig == nil {
-		return nil, errors.New("insufficient information to build KubernetesMasterConfig")
+// buildUpstreamGenericConfig copies the apiserver.Config setup code from k8s.io/kubernetes/cmd/kube-apiserver/app/server.go.
+// ONLY COMMENT OUT CODE HERE, do not modify it. Do modifications outside of this function.
+func buildUpstreamGenericConfig(s *kapiserveroptions.ServerRunOptions) (*apiserver.Config, error) {
+	// set defaults
+	if err := s.GenericServerRunOptions.DefaultAdvertiseAddress(s.SecureServing, s.InsecureServing); err != nil {
+		return nil, err
+	}
+	_, apiServerServiceIP, err := master.DefaultServiceIPRange(s.ServiceClusterIPRange)
+	if err != nil {
+		return nil, fmt.Errorf("error determining service IP ranges: %v", err)
+	}
+	if err := s.SecureServing.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String(), apiServerServiceIP); err != nil {
```

Comment:
We shouldn't have to do this.  By the time we're called, certs should be available.